File: src/metadata.ts
import {
  AsyncBuffer, ColumnChunk,
  FileMetaData, LogicalType, RowGroup,
  SchemaElement, SchemaTree, ThriftObject, TimeUnit
} from './types.ts'
import {getSchemaPath} from './schema.js'
import {deserializeTCompactProtocol} from './thrift.js'

export const defaultInitialFetchSize = 1 << 19 // 512kb

const decoder = new TextDecoder()
function decode(value: Uint8Array): string {
  return decoder.decode(value)
}

/**
 * Read parquet metadata from an async buffer.
 *
 * An AsyncBuffer is like an ArrayBuffer, but the slices are loaded
 * asynchronously, possibly over the network.
 *
 * You must provide the byteLength of the buffer, typically from a HEAD request.
 *
 * In theory, you could use suffix-range requests to fetch the end of the file,
 * and save a round trip. But in practice, this doesn't work because chrome
 * deems suffix-range requests as a not-safe-listed header, and will require
 * a pre-flight. So the byteLength is required.
 *
 * To make this efficient, we initially request the last 512kb of the file,
 * which is likely to contain the metadata. If the metadata length exceeds the
 * initial fetch, 512kb, we request the rest of the metadata from the AsyncBuffer.
 *
 * This ensures that we either make one 512kb initial request for the metadata,
 * or a second request for up to the metadata size.
 *
 * @param asyncBuffer parquet file contents
 * @param initialFetchSize initial fetch size in bytes (default 512kb)
 * @returns parquet metadata object
 */
export async function parquetMetadataAsync(asyncBuffer:AsyncBuffer, initialFetchSize: number = defaultInitialFetchSize) : Promise<FileMetaData> {
  if (!asyncBuffer || !(asyncBuffer.byteLength >= 0)) throw new Error('parquet expected AsyncBuffer')

  // fetch last bytes (footer) of the file
  const footerOffset = Math.max(0, asyncBuffer.byteLength - initialFetchSize)
  const footerBuffer = await asyncBuffer.slice(footerOffset, asyncBuffer.byteLength)

  // Check for parquet magic number "PAR1"
  const footerView = new DataView(footerBuffer)
  if (footerView.getUint32(footerBuffer.byteLength - 4, true) !== 0x31524150) {
    throw new Error('parquet file invalid (footer != PAR1)')
  }

  // Parquet files store metadata at the end of the file
  // Metadata length is 4 bytes before the last PAR1
  const metadataLength = footerView.getUint32(footerBuffer.byteLength - 8, true)
  if (metadataLength > asyncBuffer.byteLength - 8) {
    throw new Error(`parquet metadata length ${metadataLength} exceeds available buffer ${asyncBuffer.byteLength - 8}`)
  }

  // check if metadata size fits inside the initial fetch
  if (metadataLength + 8 > initialFetchSize) {
    // fetch the rest of the metadata
    const metadataOffset = asyncBuffer.byteLength - metadataLength - 8
    const metadataBuffer = await asyncBuffer.slice(metadataOffset, footerOffset)
    // combine initial fetch with the new slice
    const combinedBuffer = new ArrayBuffer(metadataLength + 8)
    const combinedView = new Uint8Array(combinedBuffer)
    combinedView.set(new Uint8Array(metadataBuffer))
    combinedView.set(new Uint8Array(footerBuffer), footerOffset - metadataOffset)
    return parquetMetadata(combinedBuffer)
  } else {
    // parse metadata from the footer
    return parquetMetadata(footerBuffer)
  }
}

/**
 * Read parquet metadata from a buffer synchronously.
 *
 * @param arrayBuffer parquet file footer
 * @returns parquet metadata object
 */
export function parquetMetadata(arrayBuffer: ArrayBuffer): FileMetaData {
  const view = new DataView(arrayBuffer)

  // Validate footer magic number "PAR1"
  if (view.byteLength < 8) {
    throw new Error('parquet file is too short')
  }
  if (view.getUint32(view.byteLength - 4, true) !== 0x31524150) {
    throw new Error('parquet file invalid (footer != PAR1)')
  }

  // Parquet files store metadata at the end of the file
  // Metadata length is 4 bytes before the last PAR1
  const metadataLengthOffset = view.byteLength - 8
  const metadata_length = view.getUint32(metadataLengthOffset, true)
  if (metadata_length > view.byteLength - 8) {
    // {metadata}, metadata_length, PAR1
    throw new Error(`parquet metadata length ${metadata_length} exceeds available buffer ${view.byteLength - 8}`)
  }

  const metadataOffset = metadataLengthOffset - metadata_length
  const reader = { view, offset: metadataOffset }
  const metadata = deserializeTCompactProtocol(reader)

  // Parse metadata from thrift data
  const version = metadata[1] as number
  const schema: SchemaElement[] = (metadata[2] as ThriftObject).map((field): SchemaElement => {
    let fieldO = field as ThriftObject
    return {
      type: fieldO[1] as number | undefined,
      type_length: fieldO[2] as number | undefined,
      repetition_type: fieldO[3] as number | undefined,
      name: decode(fieldO[4] as Uint8Array),
      num_children: fieldO[5] as number | undefined,
      converted_type: fieldO[6] as number | undefined,
      scale: fieldO[7] as number | undefined,
      precision: fieldO[8] as number | undefined,
      logical_type: logicalType(fieldO[10]),
    };
  })
  const num_rows = metadata[3] as bigint
  const row_groups = (metadata[4] as ThriftObject).map(function (rowGroup): RowGroup {
    const rowGroupField = rowGroup as ThriftObject
    const rowGroupField1 = rowGroupField[1] as ThriftObject
    return {
      columns: rowGroupField1.map(function (column): ColumnChunk {
        const columnField = column as ThriftObject
        const columnField3 = columnField[3] as ThriftObject | undefined
        const filePath = columnField[1] as Uint8Array | undefined
        return {
          file_path: filePath && decode(filePath),
          file_offset: columnField[2] as bigint,
          meta_data: columnField3 && {
            type: columnField3[1] as number,
            path_in_schema: (columnField3[3] as Uint8Array[]).map(decode),
            codec: columnField3[4] as number,
            num_values: columnField3[5] as bigint,
            total_uncompressed_size: columnField3[6] as bigint,
            total_compressed_size: columnField3[7] as bigint,
            data_page_offset: columnField3[9] as bigint,
            index_page_offset: columnField3[10] as bigint | undefined,
            dictionary_page_offset: columnField3[11] as bigint | undefined,
          },
          offset_index_offset: columnField[4] as bigint |undefined,
          offset_index_length: columnField[5]as number |undefined,
          column_index_offset: columnField[6]as bigint |undefined,
          column_index_length: columnField[7]as number |undefined,
        };
      }),
      total_byte_size: rowGroupField[2] as bigint,
      num_rows: rowGroupField[3] as bigint,
      file_offset: rowGroupField[5] as bigint | undefined,
      total_compressed_size: rowGroupField[6] as bigint | undefined,
    };
  })
  const createdByArray = metadata[6] as Uint8Array | undefined
  const created_by = createdByArray && decode(createdByArray)

  return {
    version,
    schema,
    num_rows,
    row_groups,
    created_by,
    metadata_length,
  }
}

/**
 * Return a tree of schema elements from parquet metadata.
 *
 * @param {SchemaElement[]} schema parquet metadata object
 * @returns {SchemaTree} tree of schema elements
 */
export function parquetSchema(schema: SchemaElement[]): SchemaTree {
  return getSchemaPath(schema, [])[0]!
}

function logicalType(logicalType: any): LogicalType | undefined {
  if (logicalType?.field_1) return { type: 'STRING' }
  if (logicalType?.field_2) return { type: 'MAP' }
  if (logicalType?.field_3) return { type: 'LIST' }
  if (logicalType?.field_4) return { type: 'ENUM' }
  if (logicalType?.field_5) return {
    type: 'DECIMAL',
    scale: logicalType.field_5.field_1,
    precision: logicalType.field_5.field_2,
  }
  if (logicalType?.field_6) return { type: 'DATE' }
  if (logicalType?.field_7) return {
    type: 'TIME',
    isAdjustedToUTC: logicalType.field_7.field_1,
    unit: timeUnit(logicalType.field_7.field_2),
  }
  if (logicalType?.field_8) return {
    type: 'TIMESTAMP',
    isAdjustedToUTC: logicalType.field_8.field_1,
    unit: timeUnit(logicalType.field_8.field_2),
  }
  if (logicalType?.field_10) return {
    type: 'INTEGER',
    bitWidth: logicalType.field_10.field_1,
    isSigned: logicalType.field_10.field_2,
  }
  if (logicalType?.field_11) return { type: 'NULL' }
  if (logicalType?.field_12) return { type: 'JSON' }
  if (logicalType?.field_13) return { type: 'BSON' }
  if (logicalType?.field_14) return { type: 'UUID' }
  if (logicalType?.field_15) return { type: 'FLOAT16' }
  if (logicalType?.field_16) return { type: 'VARIANT' }
  if (logicalType?.field_17) return undefined
  if (logicalType?.field_18) return undefined
  return logicalType
}

function timeUnit(unit: any): TimeUnit {
  if (unit.field_1) return TimeUnit.MILLIS
  if (unit.field_2) return TimeUnit.MICROS
  if (unit.field_3) return TimeUnit.NANOS
  throw new Error('parquet time unit required')
}

File: src/main/kotlin/app/softwork/hyparquet/Metadata.kt
package app.softwork.hyparquet

import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.max

const val defaultInitialFetchSize = 1 shl 19 // 512kb

/**
 * Read parquet metadata from an async buffer.
 */
suspend fun parquetMetadataAsync(
    asyncBuffer: AsyncBuffer,
    initialFetchSize: Int = defaultInitialFetchSize
): FileMetaData {
    if (asyncBuffer.byteLength < 0) throw Error("parquet expected AsyncBuffer")

    // fetch last bytes (footer) of the file
    val footerOffset = max(0, asyncBuffer.byteLength - initialFetchSize)
    val footerBuffer = asyncBuffer.slice(footerOffset, asyncBuffer.byteLength)

    // Check for parquet magic number "PAR1"
    val footerView = ByteBuffer.wrap(footerBuffer).order(ByteOrder.LITTLE_ENDIAN)
    if (footerView.getInt(footerBuffer.size - 4) != 0x31524150) {
        throw Error("parquet file invalid (footer != PAR1)")
    }

    // Parquet files store metadata at the end of the file
    // Metadata length is 4 bytes before the last PAR1
    val metadataLength = footerView.getInt(footerBuffer.size - 8)
    if (metadataLength > asyncBuffer.byteLength - 8) {
        throw Error("parquet metadata length $metadataLength exceeds available buffer ${asyncBuffer.byteLength - 8}")
    }

    // check if metadata size fits inside the initial fetch
    return if (metadataLength + 8 > initialFetchSize) {
        // fetch the rest of the metadata
        val metadataOffset = asyncBuffer.byteLength - metadataLength - 8
        val metadataBuffer = asyncBuffer.slice(metadataOffset, footerOffset)
        // combine initial fetch with the new slice
        val combinedBuffer = ByteArray(metadataLength + 8)
        System.arraycopy(metadataBuffer, 0, combinedBuffer, 0, metadataBuffer.size)
        System.arraycopy(footerBuffer, 0, combinedBuffer, (footerOffset - metadataOffset).toInt(), footerBuffer.size)
        parquetMetadata(combinedBuffer)
    } else {
        // parse metadata from the footer
        parquetMetadata(footerBuffer)
    }
}

/**
 * Read parquet metadata from a buffer synchronously.
 *
 * @param arrayBuffer parquet file footer
 * @returns parquet metadata object
 */
fun parquetMetadata(arrayBuffer: ByteArray): FileMetaData {
    val view = ByteBuffer.wrap(arrayBuffer).order(ByteOrder.LITTLE_ENDIAN)

    // Validate footer magic number "PAR1"
    if (view.limit() < 8) {
        throw Error("parquet file is too short")
    }
    if (view.getInt(view.limit() - 4) != 0x31524150) {
        throw Error("parquet file invalid (footer != PAR1)")
    }

    // Parquet files store metadata at the end of the file
    // Metadata length is 4 bytes before the last PAR1
    val metadataLengthOffset = view.limit() - 8
    val metadata_length = view.getInt(metadataLengthOffset)
    if (metadata_length > view.limit() - 8) {
        throw Error("parquet metadata length $metadata_length exceeds available buffer ${view.limit() - 8}")
    }

    val metadataOffset = metadataLengthOffset - metadata_length
    val reader = DataReader(view, metadataOffset)
    val metadata = deserializeTCompactProtocol(reader)

    // Parse metadata from thrift data
    val version = (metadata[1] as? ThriftType.IntType)?.value ?: 0
    val schemaList = metadata[2] as? ThriftType.ListType
    val schema = schemaList?.value?.mapNotNull { field ->
        (field as? ThriftType.ObjectType)?.let { parseSchemaElement(it.value) }
    } ?: emptyList()
    
    val num_rows = (metadata[3] as? ThriftType.LongType)?.value ?: 0L
    val rowGroupsList = metadata[4] as? ThriftType.ListType
    val row_groups = rowGroupsList?.value?.mapNotNull { rowGroup ->
        (rowGroup as? ThriftType.ObjectType)?.let { parseRowGroup(it.value) }
    } ?: emptyList()
    
    val createdByArray = (metadata[6] as? ThriftType.ByteArrayType)?.value
    val created_by = createdByArray?.decodeToString()

    return FileMetaData(
        version = version,
        schema = schema,
        num_rows = num_rows,
        row_groups = row_groups,
        created_by = created_by,
        metadata_length = metadata_length
    )
}

private fun parseSchemaElement(field: ThriftObject): SchemaElement {
    val type = (field.getOrNull(1) as? ThriftType.IntType)?.value
    val type_length = (field.getOrNull(2) as? ThriftType.IntType)?.value
    val repetition_type = (field.getOrNull(3) as? ThriftType.IntType)?.value?.let { 
        FieldRepetitionType.values()[it]
    }
    val name = (field.getOrNull(4) as? ThriftType.ByteArrayType)?.value?.decodeToString() ?: ""
    val num_children = (field.getOrNull(5) as? ThriftType.IntType)?.value
    val converted_type = (field.getOrNull(6) as? ThriftType.IntType)?.value?.let { 
        ConvertedType.values()[it]
    }
    val scale = (field.getOrNull(7) as? ThriftType.IntType)?.value
    val precision = (field.getOrNull(8) as? ThriftType.IntType)?.value
    val logical_type = (field.getOrNull(10) as? ThriftType.ObjectType)?.let { 
        parseLogicalType(it.value)
    }

    return SchemaElement(
        type = type?.let { ParquetType.values().getOrNull(it) },
        type_length = type_length,
        repetition_type = repetition_type,
        name = name,
        num_children = num_children,
        converted_type = converted_type,
        scale = scale,
        precision = precision,
        logical_type = logical_type
    )
}

private fun parseRowGroup(rowGroupField: ThriftObject): RowGroup {
    val columnsList = rowGroupField.getOrNull(1) as? ThriftType.ListType
    val columns = columnsList?.value?.mapNotNull { column ->
        (column as? ThriftType.ObjectType)?.let { parseColumnChunk(it.value) }
    } ?: emptyList()
    
    val total_byte_size = (rowGroupField.getOrNull(2) as? ThriftType.LongType)?.value ?: 0L
    val num_rows = (rowGroupField.getOrNull(3) as? ThriftType.LongType)?.value ?: 0L
    val file_offset = (rowGroupField.getOrNull(5) as? ThriftType.LongType)?.value
    val total_compressed_size = (rowGroupField.getOrNull(6) as? ThriftType.LongType)?.value

    return RowGroup(columns, total_byte_size, num_rows, file_offset, total_compressed_size)
}

private fun parseColumnChunk(columnField: ThriftObject): ColumnChunk {
    val filePath = (columnField.getOrNull(1) as? ThriftType.ByteArrayType)?.value?.decodeToString()
    val file_offset = (columnField.getOrNull(2) as? ThriftType.LongType)?.value ?: 0L
    val meta_data = (columnField.getOrNull(3) as? ThriftType.ObjectType)?.let { parseColumnMetaData(it.value) }
    val offset_index_offset = (columnField.getOrNull(4) as? ThriftType.LongType)?.value
    val offset_index_length = (columnField.getOrNull(5) as? ThriftType.IntType)?.value
    val column_index_offset = (columnField.getOrNull(6) as? ThriftType.LongType)?.value
    val column_index_length = (columnField.getOrNull(7) as? ThriftType.IntType)?.value

    return ColumnChunk(filePath, file_offset, meta_data, offset_index_offset, offset_index_length, column_index_offset, column_index_length)
}

private fun parseColumnMetaData(columnField3: ThriftObject): ColumnMetaData {
    val type = (columnField3.getOrNull(1) as? ThriftType.IntType)?.value?.let { ParquetType.values()[it] } ?: ParquetType.BOOLEAN
    val path_in_schema = (columnField3.getOrNull(3) as? ThriftType.ListType)?.value?.mapNotNull {
        (it as? ThriftType.ByteArrayType)?.value?.decodeToString()
    } ?: emptyList()
    val codec = (columnField3.getOrNull(4) as? ThriftType.IntType)?.value?.let { CompressionCodec.values()[it] } ?: CompressionCodec.UNCOMPRESSED
    val num_values = (columnField3.getOrNull(5) as? ThriftType.LongType)?.value ?: 0L
    val total_uncompressed_size = (columnField3.getOrNull(6) as? ThriftType.LongType)?.value ?: 0L
    val total_compressed_size = (columnField3.getOrNull(7) as? ThriftType.LongType)?.value ?: 0L
    val data_page_offset = (columnField3.getOrNull(9) as? ThriftType.LongType)?.value ?: 0L
    val index_page_offset = (columnField3.getOrNull(10) as? ThriftType.LongType)?.value
    val dictionary_page_offset = (columnField3.getOrNull(11) as? ThriftType.LongType)?.value

    return ColumnMetaData(type, path_in_schema, codec, num_values, total_uncompressed_size, total_compressed_size, data_page_offset, index_page_offset, dictionary_page_offset)
}

/**
 * Return a tree of schema elements from parquet metadata.
 */
fun parquetSchema(schema: List<SchemaElement>): SchemaTree {
    return getSchemaPath(schema, emptyList())[0]
}

private fun parseLogicalType(logicalType: ThriftObject): LogicalType? {
    if (logicalType.getOrNull(1) != null) return LogicalType.STRING
    if (logicalType.getOrNull(2) != null) return LogicalType.MAP
    if (logicalType.getOrNull(3) != null) return LogicalType.LIST
    if (logicalType.getOrNull(4) != null) return LogicalType.ENUM
    if (logicalType.getOrNull(5) != null) {
        val decimal = (logicalType[5] as? ThriftType.ObjectType)?.value
        val scale = (decimal?.getOrNull(1) as? ThriftType.IntType)?.value ?: 0
        val precision = (decimal?.getOrNull(2) as? ThriftType.IntType)?.value ?: 0
        return LogicalType.DECIMAL(precision, scale)
    }
    if (logicalType.getOrNull(6) != null) return LogicalType.DATE
    if (logicalType.getOrNull(7) != null) {
        val time = (logicalType[7] as? ThriftType.ObjectType)?.value
        val isAdjustedToUTC = (time?.getOrNull(1) as? ThriftType.BooleanType)?.value ?: false
        val unit = (time?.getOrNull(2) as? ThriftType.ObjectType)?.value?.let { parseTimeUnit(it) } ?: TimeUnit.MILLIS
        return LogicalType.TIME(isAdjustedToUTC, unit)
    }
    if (logicalType.getOrNull(8) != null) {
        val timestamp = (logicalType[8] as? ThriftType.ObjectType)?.value
        val isAdjustedToUTC = (timestamp?.getOrNull(1) as? ThriftType.BooleanType)?.value ?: false
        val unit = (timestamp?.getOrNull(2) as? ThriftType.ObjectType)?.value?.let { parseTimeUnit(it) } ?: TimeUnit.MILLIS
        return LogicalType.TIMESTAMP(isAdjustedToUTC, unit)
    }
    if (logicalType.getOrNull(10) != null) {
        val integer = (logicalType[10] as? ThriftType.ObjectType)?.value
        val bitWidth = (integer?.getOrNull(1) as? ThriftType.IntType)?.value ?: 32
        val isSigned = (integer?.getOrNull(2) as? ThriftType.BooleanType)?.value ?: true
        return LogicalType.INTEGER(bitWidth, isSigned)
    }
    if (logicalType.getOrNull(11) != null) return LogicalType.NULL
    if (logicalType.getOrNull(12) != null) return LogicalType.JSON
    if (logicalType.getOrNull(13) != null) return LogicalType.BSON
    if (logicalType.getOrNull(14) != null) return LogicalType.UUID
    if (logicalType.getOrNull(15) != null) return LogicalType.FLOAT16
    if (logicalType.getOrNull(16) != null) return LogicalType.VARIANT
    throw Error("unsupported logical type")
}

private fun parseTimeUnit(unit: ThriftObject): TimeUnit {
    if (unit.getOrNull(1) != null) return TimeUnit.MILLIS
    if (unit.getOrNull(2) != null) return TimeUnit.MICROS
    if (unit.getOrNull(3) != null) return TimeUnit.NANOS
    throw Error("parquet time unit required")
}
